File: pathplannerlib/src/main/java/com/pathplanner/lib/GeometryUtil.java
package com.pathplanner.lib;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;

public class GeometryUtil {
  static double doubleLerp(double startVal, double endVal, double t) {
    return startVal + (endVal - startVal) * t;
  }

  static Rotation2d rotationLerp(Rotation2d startVal, Rotation2d endVal, double t) {
    return startVal.plus(endVal.minus(startVal).times(t));
  }

  static Translation2d translationLerp(Translation2d a, Translation2d b, double t) {
    return a.plus((b.minus(a)).times(t));
  }

  static Translation2d quadraticLerp(Translation2d a, Translation2d b, Translation2d c, double t) {
    Translation2d p0 = translationLerp(a, b, t);
    Translation2d p1 = translationLerp(b, c, t);
    return translationLerp(p0, p1, t);
  }

  static Translation2d cubicLerp(
      Translation2d a, Translation2d b, Translation2d c, Translation2d d, double t) {
    Translation2d p0 = quadraticLerp(a, b, c, t);
    Translation2d p1 = quadraticLerp(b, c, d, t);
    return translationLerp(p0, p1, t);
  }

  static Rotation2d cosineInterpolate(Rotation2d y1, Rotation2d y2, double mu) {
    double mu2 = (1 - Math.cos(mu * Math.PI)) / 2;
    return new Rotation2d(y1.getRadians() * (1 - mu2) + y2.getRadians() * mu2);
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/GeometryUtil.cpp
#include "pathplanner/lib/GeometryUtil.h"

using namespace pathplanner;

units::second_t GeometryUtil::unitLerp(units::second_t startVal, units::second_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::meters_per_second_t GeometryUtil::unitLerp(units::meters_per_second_t startVal, units::meters_per_second_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::meters_per_second_squared_t GeometryUtil::unitLerp(units::meters_per_second_squared_t startVal, units::meters_per_second_squared_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::radians_per_second_t GeometryUtil::unitLerp(units::radians_per_second_t startVal, units::radians_per_second_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::radians_per_second_squared_t GeometryUtil::unitLerp(units::radians_per_second_squared_t startVal, units::radians_per_second_squared_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::meter_t GeometryUtil::unitLerp(units::meter_t startVal, units::meter_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

units::curvature_t GeometryUtil::unitLerp(units::curvature_t startVal, units::curvature_t endVal, double t) {
   return startVal + (endVal - startVal) * t;
}

frc::Rotation2d GeometryUtil::rotationLerp(const frc::Rotation2d startVal, const frc::Rotation2d endVal, double t) {
   return startVal + ((endVal - startVal) * t);
}

frc::Translation2d GeometryUtil::translationLerp(const frc::Translation2d startVal, const frc::Translation2d endVal, double t) {
   return startVal + ((endVal - startVal) * t);
}

frc::Translation2d GeometryUtil::quadraticLerp(const frc::Translation2d a, const frc::Translation2d b, const frc::Translation2d c, double t){
   frc::Translation2d p0 = GeometryUtil::translationLerp(a, b, t);
   frc::Translation2d p1 = GeometryUtil::translationLerp(b, c, t);
   return GeometryUtil::translationLerp(p0, p1, t);
}

frc::Translation2d GeometryUtil::cubicLerp(const frc::Translation2d a, const frc::Translation2d b, const frc::Translation2d c, const frc::Translation2d d, double t){
   frc::Translation2d p0 = GeometryUtil::quadraticLerp(a, b, c, t);
   frc::Translation2d p1 = GeometryUtil::quadraticLerp(b, c, d, t);
   return GeometryUtil::translationLerp(p0, p1, t);
}

units::degree_t GeometryUtil::modulo(units::degree_t a, units::degree_t b){
   return a - (b * units::math::floor(a / b));
}

bool GeometryUtil::isFinite(units::meter_t u){
   return std::isfinite(u());
}

bool GeometryUtil::isNaN(units::meter_t u){
   return std::isnan(u());
}
